tests/test_text_generation_example.py

```diff
@@ -239,9 +239,20 @@ def _test_text_generation(
         assert results["throughput"] >= (2 - TIME_PERF_FACTOR) * baseline
 
 
-@pytest.mark.parametrize("model_name, batch_size, reuse_cache, baseline", MODELS_TO_TEST["bf16"])
-def test_text_generation_bf16(model_name: str, baseline: float, batch_size: int, reuse_cache: bool, token: str):
-    _test_text_generation(model_name, baseline, token, batch_size, reuse_cache)
+@pytest.mark.parametrize("model_name, world_size, batch_size, reuse_cache, baseline", MODELS_TO_TEST["bf16"])
+def test_text_generation_bf16(
+    model_name: str, baseline: float, world_size: int, batch_size: int, reuse_cache: bool, token: str
+):
+    deepspeed = True if world_size > 1 else False
```

Comment:
`world_size` should always be 1 for this test. Maybe it could be renamed to `test_text_generation_bf16_1x`, but the the goal of this test is to run models in bf16 on a single device. And so, in the baselines above for bf16, we can remove the 1 added for the world size since it will always be 1.

Reply:
I understand your point Regiss. 
@libinta and I were discussing this and she mentioned to change this to include the `world_size` here as well for two reasons:
- making the tuple sequence between `bf16` `fp8` and `deepspeed` closely **similar**. Right now each test has it's own sequence without these changes. 
- To eventually consolidate `bf16` and `deepspeed` into a single section and add [tensor parallel distribution](https://github.com/huggingface/optimum-habana/pull/1121/files) in there as well (possibly with another tuple value) when avail. 
   - below is a test with 2x card using `falcon40b` conducted with both  `bf16` and `deepspeed` sections, leading to the results, as an example of the intended consolidation strategy we are thinking here.  

I agree that the simplicity of current `bf16`  sequence for 1x card tests is an advantage. 
Let me know if this makes scene. 

```bash
python -m pytest tests/test_text_generation_example.py -v -s -k test_text_generation_bf16 --token $HFToken

configfile: setup.cfg
collected 21 items / 20 deselected / 1 selected

tests/test_text_generation_example.py::test_text_generation_bf16[token0-tiiuae/falcon-40b-2-1-False-45.40269869717129]

Command to test: python3 /devops/sgohari/tests/codes/ig/optimum-habana/examples/gaudi_spawn.py --use_deepspeed --world_size 2 /devops/sgohari/tests/codes/ig/optimum-habana/examples/text-generation/run_generation.py --model_name_or_path tiiuae/falcon-40b --batch_size 1 --use_kv_cache --max_new_tokens 100 --use_flash_attention --flash_attention_causal_mask --use_hpu_graphs --output_dir /tmp/tmpcdz241gm


Command to test: python3 /devops/sgohari/tests/codes/ig/optimum-habana/examples/gaudi_spawn.py --use_deepspeed --world_size 2 /devops/sgohari/tests/codes/ig/optimum-habana/examples/text-generation/run_generation.py --model_name_or_path tiiuae/falcon-40b --batch_size 1 --use_kv_cache --max_new_tokens 100 --use_flash_attention --flash_attention_causal_mask --use_hpu_graphs --output_dir /tmp/tmpcdz241gm



Stats:
--------------------------------------------------------------------------------------------------------------
Throughput (including tokenization) = 45.37525150541983 tokens/second
Number of HPU graphs                = 18
Memory allocated                    = 40.02 GB
Max memory allocated                = 40.03 GB
Total memory available              = 94.62 GB
Graph compilation duration          = 8.465567941995687 seconds
--------------------------------------------------------------------------------------------------------------

[2024-07-22 17:27:26,443] [INFO] [launch.py:348:main] Process 9731 exits successfully.
[2024-07-22 17:27:26,443] [INFO] [launch.py:348:main] Process 9730 exits successfully.
PASSED

============================================================================================= 1 passed, 20 deselected in 83.64s (0:01:23) =============================================================================================
```
```bash
python -m pytest tests/test_text_generation_example.py -v -s -k test_text_generation_deepspeed --token $HFToken

platform linux -- Python 3.10.12, pytest-7.4.4, pluggy-1.5.0 -- /usr/bin/python3
cachedir: .pytest_cache
rootdir: /devops/sgohari/tests/codes/ig/optimum-habana
configfile: setup.cfg
collected 21 items / 20 deselected / 1 selected

tests/test_text_generation_example.py::test_text_generation_deepspeed[token0-tiiuae/falcon-40b-2-1-45.40269869717129]

Command to test: python3 /devops/sgohari/tests/codes/ig/optimum-habana/examples/gaudi_spawn.py --use_deepspeed --world_size 2 /devops/sgohari/tests/codes/ig/optimum-habana/examples/text-generation/run_generation.py --model_name_or_path tiiuae/falcon-40b --batch_size 1 --use_kv_cache --max_new_tokens 100 --use_flash_attention --flash_attention_causal_mask --use_hpu_graphs --output_dir /tmp/tmp79x6rm6v


Stats:
--------------------------------------------------------------------------------------------------------------
Throughput (including tokenization) = 45.39572275823871 tokens/second
Number of HPU graphs                = 18
Memory allocated                    = 40.02 GB
Max memory allocated                = 40.03 GB
Total memory available              = 94.62 GB
Graph compilation duration          = 8.42588968500786 seconds
--------------------------------------------------------------------------------------------------------------

[2024-07-22 17:29:18,960] [INFO] [launch.py:348:main] Process 11895 exits successfully.
[2024-07-22 17:29:18,960] [INFO] [launch.py:348:main] Process 11896 exits successfully.
PASSED

============================================================================================= 1 passed, 20 deselected in 77.56s (0:01:17) =============================================================================================
```

Reply:
@regisss 
Any more thoughts here?

Reply:
I think we should keep the bf16 tests with a world size of 1, and we can add others with world size > 1.
Both paths don't cover exactly the same part of the codebase + bf16 tests on a single device are faster which is useful for testing things quickly.
So we could have `test_text_generation_bf16_1x` and `test_text_generation_bf16_Kx` to cover both.

Reply:
@regisss 
Sounds good. done: https://github.com/huggingface/optimum-habana/pull/1140/commits/e3aad6a307e37d8af042ca6a5e45a015732b02e6
I reverted the changes, and left the `world_size` as is (aka 1). 
I renamed the `test_text_generation_bf16` to `test_text_generation_bf16_1x` and `bf16` MODEL_LIST to `bf16_1x` 
Let me know how does this look now

Reply:
Yes, that looks good, thanks!